addon/-private/system/relationships/state/relationship.js

```diff
@@ -316,5 +316,49 @@ Relationship.prototype = {
   setHasLoaded(value) {
     heimdall.increment(setHasLoaded);
     this.hasLoaded = value;
-  }
+  },
+
+  /*
+    `push` for a relationship allows the store to push a JSON API Relationship
+    Object onto the relationship. The relationship will then extract and set the
+    meta, data and links of that relationship.
+
+    `push` use `updateMeta`, `updateData` and `updateLink` to update the state
+    of the relationship.
+   */
+  push(payload) {
+    heimdall.increment(push);
+
+    let hasData = false;
+    let hasLink = false;
+
+    if (payload.meta) {
+      this.updateMeta(payload.meta);
+    }
+
+    if (payload.data !== undefined) {
+      hasData = true;
+      this.updateData(payload.data);
+    }
+
+    if (payload.links && payload.links.related) {
+      let relatedLink = _normalizeLink(payload.links.related);
+      if (relatedLink && relatedLink.href) {
+        hasLink = true;
+        this.updateLink(relatedLink.href);
+      }
+    }
+
+    // Pushing new data to the relationship might contain only data or links or
+    // a combination of both. Depending on what we pushed we need to set the
+    // hasData and hasLoaded flags accordingly.
+    if (hasData) {
+      this.setHasData(true);
+      this.setHasLoaded(true);
+    } else if (hasLink) {
```

Comment:
Seems like if you push the same link twice, this will say hasLoaded is false now